lib/msprime.c

```diff
@@ -6149,6 +6157,77 @@ genic_selection_print_state(sweep_t *self, FILE *out)
  * Public API for setting simulation models.
  **************************************************************/
 
+static int
+msp_scale_recomb_map_rates(msp_t *self,
+        double convert (struct _simulation_model_t *model, double rate))
+{
+    size_t i;
+    double *rates;
+    int ret;
+    recomb_map_t new_map;
+
+    rates = malloc(recomb_map_get_size(&self->recomb_map) * sizeof(double));
+    if (rates == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+
+    recomb_map_get_rates(&self->recomb_map, rates);
+    for (i = 0; i < recomb_map_get_size(&self->recomb_map); i++) {
```

Comment:
I think we can simplify this function a lot:
```c
 for (i = 0; i < recomb_map_get_size(&self->recomb_map); i++) {
     self->recomb_map.rates[i] = convert(&self->model, self->recomb_map.rates[i]);
}
```

If breaking the abstractions and reaching between objects is too nasty for you, then you could have a method like
```c
recomb_map_change_rates(recomb_map_t *self,  double convert (void *obj, double rate))
```
which would probably work.

Doing a bunch of mallocing seems overboard to me though.

Reply:
Ya that seems fine. Would also have to change the cumulative mass in the map too, but I think that's it

Reply:
would you also pass in the model as a parameter `obj` there?

Reply:
Yeah, the idea is that we can use a generic pointer here rather than insisting it's of type model_t  or whatever. This is a slightly more general interface to make it easier to test and not quite so coupled to the rest of the code.

I *think* it'll work, but if not just use the actual type.

Reply:
That worked! I was trying to think of how to make a closure when I first tackled it because it felt weird to pass a model. This is neat.